File: ch3/3-3.c
#include <stdio.h>

void expand(char s1[], char s2[]);

int main(void)
{
    char s1[] = "a-zA-Z0-9";
    char s2[1000];

    expand(s1, s2);
    printf("%s\n", s2);
    return 0;
}

void expand(char s1[], char s2[])
{
    int k = 0;
    for (int i = 0; s1[i] != '\0'; i++) {
        if (s1[i] == '-') {
            for (int j = s1[i-1]; j <= s1[i+1]; j++, k++) {
                s2[k] = (char)j;
            }
        }
    }
    s2[k+1] = '\0';
}

File: gkr/ch3/3-3.go
package main

import "fmt"

func main() {
	s1 := "-a-zA-Z0-9-"
	s2 := make([]byte, 1000)

	expand([]byte(s1), s2)
	fmt.Printf("%s\n", string(s2))
}

func expand(s1 []byte, s2 []byte) {
	var trai bool
	var l int
	for _ = range s1 {
		l++
	}
	if rune(s1[l-1]) == '-' {
		trai = true
	}
	var k, i int
	if rune(s1[0]) == '-' {
		s2[k] = '-'
		k++
		i++
	}
	for {
		if i == len(s1)-1 {
			break
		}
		if rune(s1[i]) == '-' {
			for j := rune(s1[i-1]); j <= rune(s1[i+1]); j, k = j+1, k+1 {
				s2[k] = byte(j)
			}
		}
		i++
	}
	if trai {
		s2[k] = '-'
		k++
	}
}
